docs/storage_transformers/sharding/v1.0.rst

```diff
@@ -0,0 +1,279 @@
+
+.. code-block::
+
+    | chunk (0, 0)    | chunk (0, 1)    | chunk (1, 0)    | chunk (1, 1)    |
+    | offset | length | offset | length | offset | length | offset | length |
+    | uint64 | uint64 | uint64 | uint64 | uint64 | uint64 | uint64 | uint64 |
+
+
+Empty chunks are denoted by setting both offset and length to ``2^64 - 1``.
+The index always has the full shape of all possible chunks per shard,
+even if they are outside of the array size.
+
+The actual order of the chunk-content is not fixed and may be chosen by the implementation
+as all possible write orders are valid according to this specification and therefore can
+be read by any other implementation. When writing partial chunks into an existing shard no
+specific order of the existing chunks may be expected. Some writing strategies might be
+
+* **Fixed order**: Specify a fixed order (e.g. row-, column-major or Morton order).
+  When replacing existing chunks larger or equal sized chunks may be replaced in-place,
+  leaving unused space up to an upper limit  which might possibly be specified.
+  Please note that for regular-sized uncompressed data all chunks have the same size and
+  can therefore be replaced in-place.
+* **Append-only**: Any chunk to write is appended to the existing shard,
+  followed by an updated index.
+
+Any configuration parameters for the write strategy must not be part of the metadata document,
+they need to be configured at runtime, as this is implementation specific.
+
+
+API implementation
+------------------
+
+The section below defines an implementation of the
+:ref:`abstract-store-interface` in terms of the operations of this
+storage transformer as a ``StoreWithPartialAccess``.
+The term `underlying store` references either the next storage transformer
+in the stack or the actual store if this transformer is the last one in the
+stack. Any operations with keys not starting with ``data/root`` are simply
+relayed to the underlying store and not described explicitly.
+
+* ``get_partial_values(key_ranges) -> values``:
+  For each referenced key, request the indices from the underlying store using
+  ``get_partial_values``. For each `key`, `range` pair in in `key_ranges`,
+  check if the chunk exists by checking if the index offset and length
+  are both ``2^64 - 1``. For existing keys, request the actual chunks by
+  their ranges as read from the index using ``get_partial_values``.
+  This operation should be implemented using two ``get_partial_values``
+  operations on the underlying store, one for retrieving the indices and
+  one for retrieving existing chunks.
+
+* ``set_partial_values(key_start_values)`` :
+  For each referenced key, check if all available chunks in a shard are
+  referenced. In this case a shard can be constructed according to the
+  `Binary shard format`_ directly.
+  For all other keys, request the indices from the underlying store using
+  ``get_partial_values``. All chunks that are not updated completely and
```

Comment:
how does the `get_partial_values` function return the indices?

Reply:
The index format is described above.

Reply:
So you use (store or next transformer).get_partial_values to read the index to perform the operation of the transformer.get_partial_values function?
To me, this is confusing as they are called the same but are defined on the transformer and on the store, right?

Reply:
But I guess it makes sense :)

Reply:
> So you use (store or next transformer).get_partial_values to read the index to perform the operation of the transformer.get_partial_values function?

Exactly :+1: 

> To me, this is confusing as they are called the same but are defined on the transformer and on the store, right?

Yes, a bit, but that's also part of the beauty, as nobody cares if the "next" store is actually the store or just the next transformer. I tried to always use the term `underlying store` refering to the next transformer or store, as described above:
> The term `underlying store` references either the next storage transformer
in the stack or the actual store if this transformer is the last one in the

However, it's still confusing, but I have no idea how to improve it. Do you have a suggestion?